File: 13.py
class Solution:
    def romanToInt(self, s: str) -> int:
        if s == '' : return 0
        hashmap = {
            'I' : 1,
            'II' : 2,
            'III' : 3,
            'IV' : 4,
            'V' : 5,
            'IX' : 9,
            'X' : 10,
            'XL' : 40,
            'L' : 50,
            'XC' : 90,
            'C' : 100,
            'CD' : 400,
            'D' : 500,
            'CM' : 900,
            'M' : 1000
        }
        # 最大正向匹配
        ret = 0
        max_len = 3
        index = 0
        while index < len(s):
            temp_skip = max_len
            while temp_skip > 0:
                if s[index : index + temp_skip] in hashmap:
                    ret += hashmap[s[index : index + temp_skip]]
                    index += temp_skip
                    break
                else:
                    temp_skip -= 1
        return ret

def main():
    s = 'MCMXCIV'
    print(Solution().romanToInt(s))

if __name__ == '__main__':
    main()

File: leetcode-cpp/13.cpp
#include <iostream>
#include <string>
#include <unordered_map>

using namespace std;

class Solution {
public:
    int romanToInt(string s) {
        if (!s.size()) return 0;
        unordered_map<string, int> hashmap;
        hashmap.insert({"I",1});
        hashmap.insert({"II",2});
        hashmap.insert({"III",3});
        hashmap.insert({"IV",4});
        hashmap.insert({"V",5});
        hashmap.insert({"IX",9});
        hashmap.insert({"X",10});
        hashmap.insert({"XL",40});
        hashmap.insert({"L",50});
        hashmap.insert({"XC",90});
        hashmap.insert({"C",100});
        hashmap.insert({"CD",400});
        hashmap.insert({"D",500});
        hashmap.insert({"CM",900});
        hashmap.insert({"M",1000});
        int max_len = 3, idx = 0, ret = 0, temp;
        while (idx != s.size()){
            if (idx + max_len <= s.size())
                temp = max_len;
            else
                temp = s.size() - idx;
            while (hashmap.find(s.substr(idx, temp)) == hashmap.end()) {
                temp -= 1;
            }
            ret += hashmap[s.substr(idx, temp)];
            idx += temp;
        }
        return ret;
    }
};

int main(){
    string s = "MCMXCIV";
    Solution a;
    int ret = a.romanToInt(s);
    cout << ret <<endl;
    return 0;
}
